### ISSUE TITLE
NeurIPS '17 | Dilated Recurrent Neural Networks.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#64]
Jae Duk Seo. [Only Numpy: NIPS 2017 - Implementing Dilated Recurrent Neural Networks with Interactive Code](https://towardsdatascience.com/only-numpy-nips-2017-implementing-dilated-recurrent-neural-networks-with-interactive-code-e83abe8c9b27).

**Comment by NorbertZheng** [comment]
## Overview
So I was surfing around [NIPS 2017](https://papers.nips.cc/book/advances-in-neural-information-processing-systems-30-2017?utm_campaign=Revue%20newsletter&utm_medium=Newsletter&utm_source=Deep%20Learning%20Weekly), publication page and I found this paper, “[Dilated Recurrent Neural Networks](https://papers.nips.cc/paper/6613-dilated-recurrent-neural-networks)” from [IBM Watson Research Center](http://www.research.ibm.com/labs/watson/index.shtml). So tried to implement it, and use it to classify MNIST Digits.

**Comment by NorbertZheng** [comment]
## Dilated RNNs
The key design is **Dilated Recurrent Skip Connection**. Denote $c_{t}^{(l)}$ as the cell in layer $l$ at time $t$. The dilated skip connection can be represented as

$$
c_{t}^{(l)}=f(x_{t}^{(l)},c_{t-s^{(l)}}^{(l)}).
$$

This is similar to the regular skip connection, which can be represented as

$$
c_{t}^{(l)}=f(x_{t}^{(l)},c_{t-1}^{(l)},c_{t-s^{(l)}}^{(l)}).
$$

Now the equation is bit complicated to understand in the mathematical equation form. So the authors of the paper have provided us with some graphics.

![image](https://user-images.githubusercontent.com/37235023/218028666-9f350105-16d8-4846-a176-c0c061d8c6eb.png)
So as seen above, it is typical RNN but with multiple layers as well as skipped connections here and there.

**Comment by NorbertZheng** [comment]
## Dilated RNNs on MNIST
### Network Architecture
![image](https://user-images.githubusercontent.com/37235023/218030874-d1d396f7-cb6b-4a6f-8455-23ea80d24a0e.png)
 - Red Circle: Final Output of the Network, a $(10,)$ vector for One hot encoding of the predicted number.
 - Brown Circle: Hidden State 0 for Layer 2.
 - Lime Circle: Hidden States for Layer 2.
 - Pink Circle: Hidden State 0 for Layer 1.
 - Black Circle: Hidden State for Layer 1.
 - Blue Numbers 1,2,3,4,5: Input for each Time Stamp (Please TAKE NOTE of this since I am going to use this knowledge to explain the Training / Test Data).

Above is the network we are going to build, the first layer is a standard RNN, nothing special, however the second layer is bit different, it have skip connection with interval of 2.

**Comment by NorbertZheng** [comment]
### Code Implementation
```python
class Diluted_RNN:
    
    def __init__(self,hid_state_c,hid_state_r,  wx_c,wx_r,  wh_c,wh_r):
        self.hidden_state = np.zeros((hid_state_c,hid_state_r))
        self.wx_h = np.random.randn(wx_c,wx_r)
        self.wh_h = np.random.randn(wh_c,wh_r)

        self.input  =  np.zeros((hid_state_c-1,wx_c))
        self.layers =  np.zeros((hid_state_c-1,hid_state_r))
        self.layersA=  np.zeros((hid_state_c-1,hid_state_r))

    def feed_forward(self,Time_Stamp,input):
        limitation_time_stamp = Time_Stamp - 1
        self.input[limitation_time_stamp,:]   = input
        self.layers[limitation_time_stamp,:]  = self.input[limitation_time_stamp,:].dot(self.wx_h) + self.hidden_state[limitation_time_stamp,:].dot(self.wh_h)
        self.layersA[limitation_time_stamp,:] = arctan(self.layers[limitation_time_stamp,:])
        return self.layersA[limitation_time_stamp,:]

    def backpropagation(self,Time_Stamp,gradient):
        
        limitation_time_stamp = Time_Stamp - 1
        grad_part_1 = gradient
        grad_part_2 = np.expand_dims(d_arctan(self.layers[limitation_time_stamp,:] ),axis=0)
        grad_part_x = np.expand_dims(self.input[limitation_time_stamp,:],axis=0)
        grad_part_h = np.expand_dims(self.hidden_state[limitation_time_stamp,:],axis=0)
        
        grad_x = grad_part_x.T.dot(grad_part_1 * grad_part_2)
        grad_h = grad_part_h.T.dot(grad_part_1 * grad_part_2)

        grad_pass_x = (grad_part_1 * grad_part_2).dot(self.wx_h.T)
        grad_pass_h = (grad_part_1 * grad_part_2).dot(self.wh_h.T)
        
        self.wx_h = self.wx_h - LR_x_h * grad_x 
        self.wh_h = self.wh_h - LR_h_h * grad_h 

        return grad_pass_x,grad_pass_h
```

Above is the object oriented form of the network, as seen we only have one simple feed forward function as well as one standard back propagation function.

**Comment by NorbertZheng** [comment]
### Training Data / Test Data
![image](https://user-images.githubusercontent.com/37235023/218033393-a7fd5b7d-fd62-4f4d-9029-4fdbb210d870.png)
 - Pink Box: Input at Time Stamp 1 (Vectorized 14*14 Pixel Image)
 - Yellow Box: Input at Time Stamp 2 (Vectorized 14*14 Pixel Image)
 - Blue Box: Input at Time Stamp 3 (Vectorized 14*14 Pixel Image)
 - Purple Box: Input at Time Stamp 4 (Vectorized 14*14 Pixel Image)
 - Green Box: Input at Time Stamp 5 (Vectorized 14*14 Pixel Image)

So this is what I did, I took the original image (28*28), performed Mean Pooling with (2*2) Matrix, which gives me (14*14) image. And we will use that as our input at time stamp 1. And for Inputs at other time stamps, we will cut the original (28*28) image into 4 portions and feed it to the network. Below is the code of achieving this.

```python
current_full_image = np.reshape(training_images[image_index,:],(28,28))
current_label      = np.expand_dims(training_lables[image_index,:],axis=0)

current_1_mean    = np.reshape(block_reduce(current_full_image,(2,2),np.mean),(1,-1))
current_2_block = np.reshape(current_full_image[:14,:14],(1,-1))
current_3_block = np.reshape(current_full_image[:14,14:],(1,-1))
current_4_block = np.reshape(current_full_image[14:,:14],(1,-1))
current_5_block = np.reshape(current_full_image[14:,14:],(1,-1))
```

And with that in mind, lets take a look at other training data. Finally, the reason I did this, is simply put, I wanted to LOL.
![image](https://user-images.githubusercontent.com/37235023/218033990-c0aae09a-7532-4153-a00c-38f6c325f9db.png)
Image of 3 6 4 in the respective order.

**Comment by NorbertZheng** [comment]
### Training and Results
![image](https://user-images.githubusercontent.com/37235023/218034125-a236b9e4-a32f-4753-829b-439f8a9f0110.png)
Looking at the cost over time graph we can see, that the network started out in the 4.0 region and was able to decreased the cost to 1.7.

![image](https://user-images.githubusercontent.com/37235023/218034212-ed4c4e82-5a80-41cf-9ec1-03936b426fbc.png)
However, the number of accurate prediction was only 68/100, giving us 68% accuracy. I think the splitting the image into 4 portions was a bad idea LOL.

**Comment by NorbertZheng** [comment]
### The Whole Training Code
```python
import numpy as np,sys,os
import matplotlib.pyplot as plt
from sklearn.utils import shuffle
from skimage.measure import block_reduce
from tensorflow.examples.tutorials.mnist import input_data

np.random.randn(6789)
np.set_printoptions(precision=2,suppress=True)

def tanh(x):
    return np.tanh(x)
def d_tanh(x):
    return 1 - np.tanh(x) ** 2
def ReLu(x):
    mask = (x>0) * 1.0
    return mask *x
def d_ReLu(x):
    mask = (x>0) * 1.0
    return mask 
def log(x):
    return 1 / (1 + np.exp(-1 * x))
def d_log(x):
    return log(x) * ( 1 - log(x))
def arctan(x):
    return np.arctan(x)
def d_arctan(x):
    return 1 / (1 + x ** 2)
def softmax(x):
    shiftx = x - np.max(x)
    exp = np.exp(shiftx)
    return exp/exp.sum()
# 1. Read Data
mnist = input_data.read_data_sets("../MNIST_data/", one_hot=True).test
images,label = shuffle(mnist.images,mnist.labels)
test_image_num,training_image_num = 100,5000

# 2. Declare Hyper Parameters
LR_x_h  = 0.001
LR_h_h =  0.001

num_epoch = 500
cost_array = []
total_cost = 0 
testing_images, testing_lables =images[:test_image_num,:],label[:test_image_num,:]
training_images,training_lables =images[test_image_num:test_image_num + training_image_num,:],label[test_image_num:test_image_num + training_image_num,:]

# 3. Build Model
class Diluted_RNN:
    
    def __init__(self,hid_state_c,hid_state_r,  wx_c,wx_r,  wh_c,wh_r):
        self.hidden_state = np.zeros((hid_state_c,hid_state_r))
        self.wx_h = np.random.randn(wx_c,wx_r)
        self.wh_h = np.random.randn(wh_c,wh_r)

        self.input  =  np.zeros((hid_state_c-1,wx_c))
        self.layers =  np.zeros((hid_state_c-1,hid_state_r))
        self.layersA=  np.zeros((hid_state_c-1,hid_state_r))

    def feed_forward(self,Time_Stamp,input):
        limitation_time_stamp = Time_Stamp - 1
        self.input[limitation_time_stamp,:]   = input
        self.layers[limitation_time_stamp,:]  = self.input[limitation_time_stamp,:].dot(self.wx_h) + self.hidden_state[limitation_time_stamp,:].dot(self.wh_h)
        self.layersA[limitation_time_stamp,:] = arctan(self.layers[limitation_time_stamp,:])
        return self.layersA[limitation_time_stamp,:]

    def backpropagation(self,Time_Stamp,gradient):
        
        limitation_time_stamp = Time_Stamp - 1
        grad_part_1 = gradient
        grad_part_2 = np.expand_dims(d_arctan(self.layers[limitation_time_stamp,:] ),axis=0)
        grad_part_x = np.expand_dims(self.input[limitation_time_stamp,:],axis=0)
        grad_part_h = np.expand_dims(self.hidden_state[limitation_time_stamp,:],axis=0)
        
        grad_x = grad_part_x.T.dot(grad_part_1 * grad_part_2)
        grad_h = grad_part_h.T.dot(grad_part_1 * grad_part_2)

        grad_pass_x = (grad_part_1 * grad_part_2).dot(self.wx_h.T)
        grad_pass_h = (grad_part_1 * grad_part_2).dot(self.wh_h.T)
        
        self.wx_h = self.wx_h - LR_x_h * grad_x 
        self.wh_h = self.wh_h - LR_h_h * grad_h 

        return grad_pass_x,grad_pass_h

# 4. Make the object
layer1 = Diluted_RNN(6,128,  196,128,  128,128)
layer2 = Diluted_RNN(4,10,  128, 10,  10,10)

# Func: Train all of the models
for iter in range(num_epoch):
    
    for image_index in range(len(training_images)):
        
        current_full_image = np.reshape(training_images[image_index,:],(28,28))
        current_label      = np.expand_dims(training_lables[image_index,:],axis=0)

        current_1_mean    = np.reshape(block_reduce(current_full_image,(2,2),np.mean),(1,-1))
        current_2_block = np.reshape(current_full_image[:14,:14],(1,-1))
        current_3_block = np.reshape(current_full_image[:14,14:],(1,-1))
        current_4_block = np.reshape(current_full_image[14:,:14],(1,-1))
        current_5_block = np.reshape(current_full_image[14:,14:],(1,-1))
        
        l1_1 = layer1.feed_forward(1,current_1_mean)
        _    = layer2.feed_forward(1,l1_1)

        l2_1 = layer1.feed_forward(2,current_2_block)

        l3_1 = layer1.feed_forward(3,current_3_block)
        _    = layer2.feed_forward(2,l3_1)

        l4_1 = layer1.feed_forward(4,current_4_block)

        l5_1 = layer1.feed_forward(5,current_5_block)
        output=layer2.feed_forward(3,l5_1)

        output_Soft = softmax(output)
        cost = ( -1 * (current_label * np.log(output_Soft)  + ( 1- current_label) * np.log(1 - output_Soft)  )).sum() 
        print("Current Iter :",iter,"  Current Image Index:  ",image_index ," Real Time Update Cost: ", cost,end='\r')
        total_cost+= cost

        gradient_3_2_x,gradient_3_2_h = layer2.backpropagation(3,output_Soft-current_label)
        _,             gradient_5_1_h = layer1.backpropagation(5,gradient_3_2_x)

        _,             gradient_4_1_h = layer1.backpropagation(4,gradient_5_1_h)

        gradient_2_2_x,gradient_2_2_h = layer2.backpropagation(2,gradient_3_2_h)
        _,             gradient_3_1_h = layer1.backpropagation(3,gradient_2_2_x+gradient_4_1_h)

        _,             gradient_2_1_h = layer1.backpropagation(2,gradient_3_1_h)

        gradient_1_2_x,_              = layer2.backpropagation(1,gradient_2_2_h)
        _,             _              = layer1.backpropagation(1,gradient_1_2_x+gradient_2_1_h)
        
    if iter % 2 == 0 :
        print('\n======================================')
        print("current Iter: ", iter, " Current Total Cost :", total_cost/len(training_images))

        for current_batch_index in range(5):
    
            testing_images,testing_lables = shuffle(testing_images,testing_lables)

            current_full_image = np.reshape(testing_images[current_batch_index,:],(28,28))
            current_label      = np.expand_dims(testing_lables[current_batch_index,:],axis=0)

            current_1_mean    = np.reshape(block_reduce(current_full_image,(2,2),np.mean),(1,-1))
            current_2_block = np.reshape(current_full_image[:14,:14],(1,-1))
            current_3_block = np.reshape(current_full_image[:14,14:],(1,-1))
            current_4_block = np.reshape(current_full_image[14:,:14],(1,-1))
            current_5_block = np.reshape(current_full_image[14:,14:],(1,-1))

            l1_1 = layer1.feed_forward(1,current_1_mean)
            _    = layer2.feed_forward(1,l1_1)

            l2_1 = layer1.feed_forward(2,current_2_block)

            l3_1 = layer1.feed_forward(3,current_3_block)
            _    = layer2.feed_forward(2,l3_1)

            l4_1 = layer1.feed_forward(4,current_4_block)

            l5_1 = layer1.feed_forward(5,current_5_block)
            output=layer2.feed_forward(3,l5_1)

            output_Soft = softmax(output)


            print('Mid Training Test Current Predict : ',
            np.where(output_Soft == output_Soft.max())[0], 
            "    Ground Truth : ",
            np.where(current_label[0] == current_label[0].max())[0] ,'\n',
            "Array : ",output_Soft,'\n'
            "Array: ",current_label)
        print('======================================')

    cost_array.append(total_cost/len(training_images))
    total_cost = 0       


print('=======================================')
print('==============FINAL====================')
correct = 0
for current_batch_index in range(len(testing_images)):

    current_full_image = np.reshape(testing_images[current_batch_index,:],(28,28))
    current_batch_label = np.expand_dims(testing_lables[current_batch_index,:],axis=0)

    current_1_mean    = np.reshape(block_reduce(current_full_image,(2,2),np.mean),(1,-1))
    current_2_block = np.reshape(current_full_image[:14,:14],(1,-1))
    current_3_block = np.reshape(current_full_image[:14,14:],(1,-1))
    current_4_block = np.reshape(current_full_image[14:,:14],(1,-1))
    current_5_block = np.reshape(current_full_image[14:,14:],(1,-1))

    l1_1 = layer1.feed_forward(1,current_1_mean)
    _    = layer2.feed_forward(1,l1_1)

    l2_1 = layer1.feed_forward(2,current_2_block)

    l3_1 = layer1.feed_forward(3,current_3_block)
    _    = layer2.feed_forward(2,l3_1)

    l4_1 = layer1.feed_forward(4,current_4_block)

    l5_1 = layer1.feed_forward(5,current_5_block)
    output=layer2.feed_forward(3,l5_1)

    output_Soft = softmax(output)

    print(' Current Predict : ',np.where(output_Soft == output_Soft.max())[0], 
    " Ground Truth : ", np.where(current_batch_label[0] == current_batch_label[0].max())[0] )

    if np.where(output_Soft == output_Soft.max())[0] == np.where(current_batch_label[0] == current_batch_label[0].max())[0]:
        correct += 1

print('Correct : ',correct, ' Out of : ',len(testing_images) )
plt.title("Cost over time")
plt.plot(np.arange(len(cost_array)), cost_array)
plt.show()

# -- end code --
```

**Comment by NorbertZheng** [comment]
## References
 - Chang, S., Zhang, Y., Han, W., Yu, M., Guo, X., Tan, W., … & Huang, T. S. (2017). Dilated recurrent neural networks. In Advances in Neural Information Processing Systems (pp. 76–86).
 - [1]2018. [Online]. Available: https://www.quora.com/What-is-the-benefit-of-using-average-pooling-rather-than-max-pooling. [Accessed: 07- Feb- 2018].
 - [3]S. Chang, Y. Zhang, W. Han, M. Yu, X. Guo, W. Tan, X. Cui, M. Witbrock, M. Hasegawa-Johnson and T. Huang, “Dilated Recurrent Neural Networks”, Papers.nips.cc, 2018. [Online]. Available: https://papers.nips.cc/paper/6613-dilated-recurrent-neural-networks. [Accessed: 14- Feb- 2018].
 - IBM Thomas J. Watson Research Center. (n.d.). Retrieved February 14, 2018, from http://www.research.ibm.com/labs/watson/index.shtml.
 - Official Code Repository. https://github.com/code-terminator/DilatedRNN.
